File: C++/maximum_score_after_splitting_a_string.cpp
// no. 1422
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
  int maxScore(string s) {
    int max_score = 0;
    vector<int> ones;
    for (int i = 0; i < s.length() - 1; i++) {
      if (s[i] == '1') {
        ones.push_back(i);
      }
    }
    ones.push_back(s.length() - 1);
    for (int i = 0; i < ones.size(); i++) {
      int j = ones[i];
      if (j == 0) {
        continue;
      }
      int right_score = ones.size() - i;
      if (s.back() == '0') {
        right_score--;
      }
      int left_score = j - i;
      int total = left_score + right_score;
      if (total > max_score) {
        max_score = total;
      }
    }
    return max_score;
  }
};

File: go/maximum_score_after_splitting_a_string.go
// no. 1422
func maxScore(s string) int {
	max_score := 0
	var ones []int
	for i := 0; i < len(s)-1; i++ {
		if s[i] == '1' {
			ones = append(ones, i)
		}
	}
	ones = append(ones, len(s)-1)
	for i := 0; i < len(ones); i++ {
		j := ones[i]
		if j == 0 {
			continue
		}
		right_score := len(ones) - i
		if s[len(s)-1] == '0' {
			right_score--
		}
		left_score := j - i
		total := left_score + right_score
		if total > max_score {
			max_score = total
		}
	}
	return max_score
}
